File: 114. Flatten Binary Tree to Linked List.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {void} Do not return anything, modify root in-place instead.
 */
var flatten = function(root) { // Recursively flatten subleft and subright trees, and then concat root and the flatten sub trees
    if (root) {
        const flattenLeftTree = flatten(root.left);
        const flattenRightTree = flatten(root.right);
        root.left = null;
        if (flattenLeftTree) {
            let previous = flattenLeftTree.right;
            while(previous.right) {       // Note, we need to get the most right leaf node, and then set its right to the root node's flatten right subtree
                previous = previous.right; 
            }
            previous.right = flattenRightTree;
            root.right = flattenLeftTree;
        } else {
            root.right = flattenRightTree;
        }
    }
    return root;
};

File: Golang/Tree/114.flatten-binary-tree-to-linked-list.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func flatten(root *TreeNode) {
	if root != nil {
		if root.Left != nil {
			prev := root.Left
			for prev.Right != nil {
				prev = prev.Right
			}
			prev.Right = root.Right
			root.Right = root.Left
			root.Left = nil
		}
		flatten(root.Right)
	}
}
